rotkehlchen/tasks/historical_balances.py

```diff
@@ -0,0 +1,224 @@
+PROTOCOL_BUCKET_IN_SUBTYPES: Final = (
+    HistoryEventSubType.RECEIVE_WRAPPED,
+    HistoryEventSubType.GENERATE_DEBT,
+)
+PROTOCOL_BUCKET_OUT_SUBTYPES: Final = (
+    HistoryEventSubType.RETURN_WRAPPED,
+    HistoryEventSubType.PAYBACK_DEBT,
+)
+
+
+def get_bucket_for_event(event: 'HistoryBaseEntry') -> tuple[str, str | None, str | None, str]:
+    """Returns (location, location_label, bucket_protocol, asset) bucket.
+
+    Most events affect the wallet bucket (protocol=NULL).
+    Only wrapped token receipts/returns and debt generation/payback
+    affect protocol-specific buckets.
+    """
+    location = event.location.serialize_for_db()
+    asset = event.asset.identifier
+    if (
+        (
+            (counterparty := getattr(event, 'counterparty', None)) is not None and
+            (direction := event.maybe_get_direction()) is not None
+        ) and
+        (
+            (event.event_subtype in PROTOCOL_BUCKET_IN_SUBTYPES and direction == EventDirection.IN) or  # noqa: E501
+            (event.event_subtype in PROTOCOL_BUCKET_OUT_SUBTYPES and direction == EventDirection.OUT)  # noqa: E501
+        )
+    ):
+        return (location, event.location_label, counterparty, asset)
+
+    return (location, event.location_label, None, asset)
+
+
+def _load_bucket_balances_before_ts(
+        database: 'DBHandler',
+        from_ts: Timestamp,
+) -> dict[tuple[str, str | None, str | None, str], FVal]:
+    """Load the latest balance per bucket from event_metrics for events before from_ts.
+
+    For each bucket (location, location_label, protocol, asset), we need the balance
+    from the last event before from_ts. Query is ordered by timestamp DESC so first
+    occurrence per bucket is the latest balance.
+    """
+    bucket_balances: dict[tuple[str, str | None, str | None, str], FVal] = {}
+    with database.conn.read_ctx() as cursor:
+        cursor.execute(
+            """
+            SELECT he.location, he.location_label, em.protocol, he.asset, em.metric_value
+            FROM event_metrics em
+            INNER JOIN history_events he ON em.event_identifier = he.identifier
+            WHERE em.metric_key = 'balance' AND he.timestamp < ?
+            ORDER BY he.timestamp DESC, he.sequence_index DESC
+            """,
+            (from_ts,),
+        )
+        for row in cursor:
+            bucket = (row[0], row[1], row[2], row[3])
+            if bucket not in bucket_balances:  # first occurrence = latest
+                bucket_balances[bucket] = FVal(row[4])
+
+    log.debug(f'Loaded {len(bucket_balances)} bucket balances before ts={from_ts}')
+    return bucket_balances
+
+
+def process_historical_balances(
+        database: 'DBHandler',
+        msg_aggregator: 'MessagesAggregator',
+        from_ts: Timestamp | None = None,
+) -> None:
+    """Process events and compute balance metrics. Stops on negative balance."""
+    log.debug(f'Starting historical balance processing from_ts={from_ts}')
```

Comment:
this stops completly on negative balances but sometimes is annoying because is small amounts for assets that you dont care. i believe is something common. i believe you shouldnt stop the whole process and just notify about the asset and continue.

Maybe ask lefteris for feedback. Not sure if this was discussed already